docs/designs/allreduce.md

```diff
@@ -1,14 +1,215 @@
+        elif res == FAILED:
+            continue
+```
+
+We also provide API to broadcast all model parameters from one of the worker pods to all other active worker pods.
+This would be useful to initialize the model parameters for a restarted or new joining worker pod. The interface would
+look like the following:
+
+```python
+communicator = BroadcastCommunicator()
+init_local_model(worker_0_ip)
+communicator.broadcast(worker_ips, get_model_params(worker_0_ip)
+```
+
+### Allreduce-based Training in ElasticDL
+
+#### Gradients Averaging and Model Updating
+
+For training based on parameter servers, gradients calculation and model updating include the following steps:
+
+1. Send the gradients calculated locally from each worker to parameter servers.
+1. Calculate the average of all the received gradients on master.
+1. Update the model in master.
+
+On contrary, in allreduce-based training, each worker in ElasticDL calculates gradients locally and then calculates
+the average of gradients across all workers using collective communication via `AllreduceCommunicator.average_gradients()`
+that we mentioned in the previous section. The main differences are the following:
+
+1. Gradients from each worker are not sent to master.
+1. The average of gradients across all workers is calculated locally on each worker.
+1. The model is updated directly on each worker and each worker has the exact same copy of the model.
+
+Below is the pseudo-code for this process on each worker:
+
+```python
+communicator = AllreduceCommunicator()
+
+with tf.GradientTape() as tape:
+    outputs = model(features, training=True)
+    loss = loss_fn(outputs, labels)
+local_grads = tape.gradient(loss, get_trainable_items())
+res, averaged_grads = communicator.average_gradients(grads)
+if res == SUCCESS:
+    update_model(averaged_grads)
+else:
+    report_failure()
+    continue
+```
+
+#### Failure Handling during Training
+
+The above pseudo-code will be wrapped and executed for each batch of the dataset inside `process_minibatch_and_report()`
+in the code below. Each worker continues to perform tasks until there is no new batch available.
+
+```python
+while True:
+    dataset = self._task_data_service.get_dataset()
+    if not dataset:
+        break
+    dataset = dataset.batch(self._minibatch_size).prefetch(1)
+    for dataset_batch in dataset:
+        task = self._task_data_service.get_current_task()
+        process_minibatch_and_report(dataset_batch, task)
+```
+
+If any of the gradients averaging operation fails while the workers are still active and healthy, we simply report the
+failure and continue training on the next batch.
+
+If any of the workers fails, e.g. the pod is accidentally killed, the following steps will be performed:
+
+1. The task that the failed worker was handling will be added back to the task queue.
+1. Since the `AllreduceCommunicator` is aware of the failure. It will reconstruct the communicator and re-assign ranks among
```

Comment:
If one worker pod is killed and restart, the communicator need be reconstructed twice?